crates/sparrow-scheduler/src/task.rs

```diff
@@ -81,10 +78,21 @@ impl Task {
     ) -> error_stack::Result<bool, Error> {
         let guard = self.schedule_count.guard();
         self.pipeline()?
-            .do_work(self.partition, scheduler)
+            .do_work(self.pending.partition, scheduler)
             .change_context_lazy(|| self.error("do_work"))?;
         Ok(guard.finish())
     }
+
+    /// Mark this task as completed.
+    ///
+    /// After this, it should not be scheduled nor should work be done.
+    pub fn complete(&self) {
+        self.pending.complete()
```

Comment:
This should let the existing operations be simplified, since the task now tracks completion (previously each operation tracked completion itself).